### PR TITLE
Phase 6/pluggable rules

### PR DESCRIPTION
**Pull request by DiyarY**
- In game of life Phase 6 the main task consists to implement the standard- and additional-rules <br> and to to each implemented rule a unit-test must be defined as well.

- Here an link to get familiar with the respective rules: [ConwaysGameOfLIfe](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life)

- At the end it's important that every class still works as expected and additional to check
that every function is beeing used and used correctly, which can be controlled with the following<br> codecoverage down below:

Screenshot (80)

Screenshot (81)

### CODE REVIEW

File: TestProject/src/GameLogic.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
+<?php
+namespace GameOfLife;
+
+use GameOfLife\Rules\BaseRule;
+use GameOfLife\cellularAutomat\Board;
+
+/**
+ * Handles the logic of the game which must be set during the before the execution process.
```
Comment (Shyru) [inline TestProject/src/GameLogic.php]: during or before?

File: TestProject/src/GameLogic.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    public function __construct(BaseRule $_rule)
+    {
+        $this->rule = $_rule;
+    }
+
+    /**
+     * Calculates the next generation of the board based on the applied rule implementation.
+     *
+     * @param Board $_board current board.
```
Comment (Shyru) [inline TestProject/src/GameLogic.php]: Please document the return value! (Explain what is returned)

File: TestProject/src/cellularAutomat/Board.php [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+ *   4. Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
+ */
+class Board
+{
+    /**
+     * @var $height
+     * @var $width
+     * @var $board = [[]]
+     * @var array[]|Field[][]
+     */
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: 1. If you document members, you must document each and everyone on itself before the actual line.
2. There is no benefit of just writing `@var $height` and then nothing afterwards it. I has 0 information in it. No type is documented, no explanation is given what it means. So, if you think a variable deserves no documentation, then remove this useless documentation block.

File: TestProject/src/cellularAutomat/Board.php [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    function __construct($_height, $_width)
+    {
+        $this->height=$_height;
+        $this->width=$_width;
+        $this->board = $this->emptyField();
+    }
+
+    /**
+     * Returns an empty Board.
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: That is not true, it does not return a Board, it returns an array with Field objects.

File: TestProject/src/cellularAutomat/Board.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+     *
+     * @return Field[][] Represents a field.
+     */
+    public function emptyField(): array
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: Why is this method public? Should't this be a private method that is only used by the board itself? Or is it used from the outside also? Is this really necessary?

File: TestProject/src/cellularAutomat/Board.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        return $newBoard;
+    }
+
+    /**
+     * Returns height of board.
+     *
+     * @return array[]|Field[][]|int
+     */
+    function getHeight()
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: As per coding style this must be named `height()`.
Remember to use the PHPStorm refactoring tools so that all usages of it are automatically adjusted.

File: TestProject/src/cellularAutomat/Board.php [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    {
+        return $this->height;
+    }
+
+    /**
+     * Returns width of the board.
+     *
+     * @return int
+     */
+    function getWidth(): int
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: As per coding style this must be named `width()`

File: TestProject/src/cellularAutomat/Board.php [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    /**
+     * Changes the value of a cell.
+     *
+     * @param int $_x X-coordinate
+     * @param int $_y Y-coordinate
+     * @param bool $_value State of the cell.
+     * @return mixed
+     */
+    function setCell($_x, $_y, bool $_value)
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: Is this private or public?

File: TestProject/src/cellularAutomat/Board.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        return $_value;
+    }
+
+    /**
+     * Returns the number of the living neighbours around the set cell.
+     *
+     * @param Field $_field The field to get the neighbors from
+     * @return Field[] All the neighbors of the given \c $_field
+     */
+    function getNeighborsOfField(Field $_field)
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: Is this private or public?

File: TestProject/src/cellularAutomat/Board.php [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+        return $this->board[$_x][$_y];
+    }
+
+    /**
+     * Returns a copy of the board.
+     *
+     * @return array Returns board.
+     */
+    public function getFieldBoard(): array
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Board.php]: Where is this method used? Is it really necessary?

File: TestProject/src/cellularAutomat/Field.php [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        $this->isAlive = $_value;
+    }
+
+    /**
+     * Returns the new value of the customised cell.
+     *
+     * @return bool new value of the cell.
+     */
+    public function getValue()
```
Comment (Shyru) [inline TestProject/src/cellularAutomat/Field.php]: As per the coding style, this should be named `value()`.

File: TestProject/src/gameoflife.php [hunk 12]
```diff
@@ -153,18 +165,39 @@ function myAutoload($className) {
...
         $output->startOutput($options);
     }
-    else die ("Requested output $requestedOutput doesn't inherit from output\\BaseOutput!\m");
+    else die ("Requested output $requestedOutput doesn't inherit from output\\BaseOutput!\n");
 }
 else die ("Could not find output $requestedOutput!\n");
 
 
+//Initialize and returns the defined rule-class inside the rule-directory.
+$requestedRule = $options->getOption("StandardRule") ?? "CopyRule";
```
Comment (Shyru) [inline TestProject/src/gameoflife.php]: In my opinion this should read `$requestedRule = $options->getOption("rule") ?? "StandardRule";`
Please check if the default rule is the standard rule if none is specified, and also test that it is possible to define other rules with the `--rule` parameter.

File: TestProject/src/rules/BaseRule.php [hunk 13]
```diff
@@ -0,0 +1,39 @@
+<?php
+namespace GameOfLife\Rules;
+
+use GameOfLife\options\Getopt;
+use GameOfLife\cellularAutomat\Field;
+
+/**
+ * API-class for additional rules.
```
Comment (Shyru) [inline TestProject/src/rules/BaseRule.php]: API-class is not a good name in this case.
Better description would be: `Base Rule for all rule implementations`.
For baseclasses it is also important to explain what a developer should do when implementing it. This is also helpful in the method documentation of the methods that should be implemented.

File: TestProject/src/rules/BaseRule.php [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+ */
+abstract class BaseRule
+{
+    /**
+     * @param Field $_field The cell which is going to be checked.
```
Comment (Shyru) [inline TestProject/src/rules/BaseRule.php]: Here you could place a hint for the developer and explain how the new state could be calculated. For example: What methods does the Field provide that are helpful to implement that?

File: TestProject/src/rules/BaseRule.php [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+     * @return bool The new state of the given field in the next generation
+     */
+    public function calculateNewState(Field $_field)
+    {
+    }
```
Comment (Shyru) [inline TestProject/src/rules/BaseRule.php]: Since there is no default implementation for this, you should define the class as abstract and remove the { and }.

File: TestProject/src/rules/BaseRule.php [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+     * @param Getopt $_options option list.
+     * @return void
+     */
+    public function addOptions(Getopt $_options)
+    {
+    }
+
+    /**
+     * Initializes the rule.
+     *
```
Comment (Shyru) [inline TestProject/src/rules/BaseRule.php]: Please explain when this is called and what it can be used for, so that the developer implementing a rule knows what to do with it!

File: TestProject/tests/GameLogicTest.php [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+        $boardLayout = <<<BOARD
+        ***
+        ---
+        ---
+BOARD;
+
+        $boardFiller = new BoardFromTextFiller();
+        $boardFiller->fillBoardFromText($board, $boardLayout);
+        $consoleOutput = new ConsoleOutput();
+        $consoleOutput->outputBoard($board);
```
Comment (Shyru) [inline TestProject/tests/GameLogicTest.php]: Please comment this out. Unit-Tests should be silent and not produce any output. For debugging purposes it is ok to temporarily add output, but if everything works as expected you should remove it or comment it.
**Please also do this for all other occurrences in all tests!**

File: TestProject/tests/GameLogicTest.php [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+        for ($i = 0; $i < 5; $i++)
+        {
+            $nextGeneration = $gameLogic->calculateNextGeneration($board);
+            $nextGeneration = $gameLogic->calculateNextGeneration($nextGeneration);
```
Comment (Shyru) [inline TestProject/tests/GameLogicTest.php]: Why do you calculate two generations per loop?

File: TestProject/tests/GameLogicTest.php [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+            if ($nextGeneration === $temp)
+            {
+                $this->assertTrue($gameLogic->isLoopDetected());
+                break;
+            }
+            else $this->assertFalse($gameLogic->isLoopDetected());
```
Comment (Shyru) [inline TestProject/tests/GameLogicTest.php]: I don't understand this, please explain why this is `if` is needed.

File: TestProject/tests/cellularAutomat/BoardTest.php [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     * Asserts that the created board  is empty.
+     */
+    public function testBoardIsEmpty()
+    {
+        $board = $this->board->getFieldBoard();
+        $boardIsEmpty = $this->checkBoardArrayEmpty($board);
+
+        $this->assertTrue($boardIsEmpty);
```
Comment (Shyru) [inline TestProject/tests/cellularAutomat/BoardTest.php]: Please combine this into one line, no need to define a variable and only use it once.

File: TestProject/tests/cellularAutomat/BoardTest.php [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+     * Asserts that the created board is not empty.
+     */
+    public function testCreatesNonEmptyBoardBySettingACellOnTrue()
+    {
+        $this->board->setCell(0,0,1);
+        $board = $this->board->getFieldBoard();
+
+        $boardIsEmpty = $this->checkBoardArrayEmpty($board);
+
+        $this->assertNotTrue($boardIsEmpty);
```
Comment (Shyru) [inline TestProject/tests/cellularAutomat/BoardTest.php]: Please combine also.

File: TestProject/tests/input/GliderTest.php [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+        $board = new Board(5, 5);
+
+        $gliderArray = [
+            [0, 0, 0, 0, 0],
+            [0, 0, 0, 0, 0],
+            [0, 0, 0, 0, 1],
+            [0, 0, 1, 0, 1],
+            [0, 0, 0, 1, 1]
+        ];
```
Comment (Shyru) [inline TestProject/tests/input/GliderTest.php]: I think this test is broken!
Why is the glider in the lower right, if you set the coordinates to 1,1?
I think your mocking code is wrong. Should'nt you mock the method `getOption`? Because that is the method that will be used from the input to determine the position, or not?

File: TestProject/tests/input/RandomTest.php [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * Asserts that the board, with a defined filling-level that sets the random-chosen cells on true, is not empty.
+     */
+    public function testFillBoardAndSetFillGrade()
+    {
+        $board = new Board(10, 10);
+        $this->getOption->method("addOptions")
+            ->with("fillingLevel")
+            ->willReturn("50");
```
Comment (Shyru) [inline TestProject/tests/input/RandomTest.php]: Same here, a better test would have been to fill it only with 10%, because 50% is the default anyways. :-)

File: TestProject/tests/output/ConsoleOutputTest.php [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Asserts that two strings, which are representing the current board, are equal.
+     *
+     * @test
+     */
+    public function outputOfAnEmptyBoard()
+    {
+        $this->expectOutputString("-----\n-----\n-----\n-----\n-----\n\n");
+        $this->boardOutput->outputBoard($this->board);
```
Comment (Shyru) [inline TestProject/tests/output/ConsoleOutputTest.php]: Nice that you found out how to test the output with phpunit! :clap:

File: TestProject/tests/output/PNGOutputTest.php [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+        $this->pngOutput->startOutput($this->getOpt);
+        $this->pngOutput->outputBoard($this->board);
+        $this->pngOutput->finishOutput();
+
+        $pngImage = imagecreate(5, 5);
+        imagecolorallocate($pngImage, 0, 0, 0);
+        imagecolorallocate($pngImage, 255, 255, 255);
+        imagepng($pngImage, "imageOutput/referenceImage.png");
+
+        $currentPngImage = file_get_contents("imageOutput/pngImage-000.png");
```
Comment (Shyru) [inline TestProject/tests/output/PNGOutputTest.php]: This is wrong, you should load the reference image here! And also rename the variable to `$referenceImage` so that it is more clear what it is.

File: TestProject/tests/output/PNGOutputTest.php [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+        $this->assertEquals($currentPngImage, file_get_contents("imageOutput/pngImage-000.png"));
+    }
+}
```
Comment (Shyru) [inline TestProject/tests/output/PNGOutputTest.php:85]: Please also add a test with a board that is non-empty.
To make it easier: It is ok that you save a reference image from the current output of the png-output, and commit it to the repository (somewhere inside the tests directory) and compare the generated image against the reference image. This at least allows you to detect if the output of the png changes in the future.

**Review by Shyru — CHANGES_REQUESTED** [review]
Good work! It took a long time, but the quality is good!
Most needed changes are concerned regarding documentation or tests, that is good sign!
However I wonder why the code coverage is so low, after having read all the tests. Did you dig a little bit deeper into the input/output directories to look why the code coverage is so low?
Please add screenshots of code-coverage output of these directories to the pull-request description so that I can have a look at it in the next round!
